README.md
CPS2Emu port for Raspberry PI
=============================

This is a port of CPS2Emu for the Raspberry PI.

How to compile
--------------

### On the Raspbery Pi

You need sdl development packages. For Raspbian:

    sudo apt-get install libsdl-dev

Then

    make

### Cross compile

To cross compile you have to set the environment variables CROSS_COMPILE to the prefix of the toolchain and SYS_ROOT to the root of a Raspbian image. 
I recommend using [this](https://github.com/raspberrypi/tools) toolchain and to
follow [this](http://www.raspberrypi.org/phpBB3/viewtopic.php?f=31&t=8478) to get a Raspbian image as SYS_ROOT.
After that simply run

    make

How to run
----------

After compiling you'll get two executables: cps2emu and cps2romcnv. The first one is the emulator and the latter is a tool to cache gfx from roms (I don't know what it does in full detail).
Create a cps2emu directory with the following structure:
- config/ -- This is the config directory from this repository, copy it.
- roms/ -- This will hold the roms.
- cps2emu
- cps2romcnv

Place the roms zip in the roms/ directory and from the csp2emu directory run

    ./cps2romcnv <rom>

where <rom> is the rom name you want to run (without the extension).
This will create the cache file for the rom, which is mandatory. It can take a while to run, so be patient.
After that you can run

    ./cps2emu <rom> #without the extension

And that's it.

Keys
----

Arrows: stick

ASDZCX: buttons

T: insert coin

U: start

L: menu

Issues
------

- The sound is not working
- It doesn't erase the whole screen when going from a menu to the game.


debian/changelog
cps2emu (0.6-r1) unstable; urgency=low
  * postinst revised, again
  * We not need ~/MyDocs/ROMs/CPS2 anymore
  * State SAVE and LOAD working, use 'l' to show MENU
  * CPS2gui now can load from any folder
  * Last file and folder used, when reopen cps2gui.
  * HiScore working, romname.hi are saved to "./hi"
  * Save state on exit and load state on start a game, saved to "./state"
  * Autocreate ./nvram and ./hi
  * Config file at ~/.config/cps2emu/cps2gui.conf

 -- Marcus Fazzi (Anunakin) <anunakin@ieee.org>  Thu, 28 Oct 2010 15:56:13 -0300

cps2emu (0.5) unstable; urgency=high
  * postinst revised
  * first release for extras-devel
  * cps2gui package merged

 -- Marcus Fazzi (Anunakin) <anunakin@ieee.org>  Tue, 26 Oct 2010 10:11:55 -0300

cps2emu (0.4) unstable; urgency=low
  * romcnv renamed to cps2romcnv
  * cps2romcnv and cps2emu compiles with same Makefile
  * Now we have a cps2gui!
  * cps2gui, not in same Makefile and .deb, for now! ;)

 -- Marcus Fazzi (Anunakin) <anunakin@ieee.org>  Sat, 23 Oct 2010 22:38:13 -0300

cps2emu (0.3) unstable; urgency=low
  * Support to PS3 Dualshock/Sixaxis controlers.

 -- Marcus Fazzi (Anunakin) <anunakin@ieee.org>  Thu, 21 Oct 2010 09:36:45 -0300

cps2emu (0.2) unstable; urgency=low
  * XSBC-Bugtracker added to debian/control
  * extras-devel release
  * Insane ARM Cortex A8 CFLAGs!

 -- Marcus Fazzi (Anunakin) <anunakin@ieee.org>  Tue, 19 Oct 2010 14:46:15 -0300

cps2emu (0.1) unstable; urgency=high
  * Binaries and Sources from garage.maemo.org.

 -- Oriol Farras <wereoffs@gmail.com>  Mon, 18 Jan 2010 09:34:34 -0300



src/draw_arm.S
	;@ sprite drawing

#include "draw_arm.h"

.global draw_tile_8_norm
.global draw_tile_8_xyflip
.global draw_tile_8_xflip
.global draw_tile_8_yflip
.global draw_tile_16_norm
.global draw_tile_16_xyflip
.global draw_tile_16_xflip
.global draw_tile_16_yflip
.global draw_tile_rows_norm
.global draw_tile_rows_xyflip
.global draw_tile_rows_xflip
.global draw_tile_rows_yflip
.global draw_tile_32_norm
.global draw_tile_32_xyflip
.global draw_tile_32_xflip
.global draw_tile_32_yflip

	.align 4

	;;@ Draw a line of a sprite.
.macro DRAW_SPR_LINE8
	ldr regTile,[regSrc, #4]
	@cmp regTile,#0
	@beq 1f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #2]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #4]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #6]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #8]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #10]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #12]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #14]

@1:
.endm

	;;@ Draw a line of a sprite xflipped.
.macro DRAW_SPR_LINE8_XFLIP
	ldr regTile,[regSrc,#4]
	@cmp regTile,#0
	@beq 1f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #2]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #4]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #6]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #8]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #10]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #12]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #14]

@1:
.endm

.macro DRAW_LINE8_XFLIP
	DRAW_SPR_LINE8_XFLIP
	add regSrc, regSrc, #8
	add regDst, regDst, #PITCH
.endm
.macro DRAW_LINE8_XYFLIP
	DRAW_SPR_LINE8_XFLIP
	add regSrc, regSrc, #8
	sub regDst, regDst, #PITCH
.endm
.macro DRAW_LINE8_YFLIP
	DRAW_SPR_LINE8
	add regSrc, regSrc, #8
	sub regDst, regDst, #PITCH
.endm
.macro DRAW_LINE8_NORM
	DRAW_SPR_LINE8
	add regSrc, regSrc, #8
	add regDst, regDst, #PITCH
.endm

	;;@ Draw a line of a sprite.
.macro DRAW_SPR_LINE16
	ldr regTile,[regSrc]
	@cmp regTile,#0
	@beq 1f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #2]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #4]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #6]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #8]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #10]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #12]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #14]

@1:
	ldr regTile,[regSrc,#4]
	@cmp regTile,#0
	@beq 2f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #16]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #18]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #20]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #22]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #24]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #26]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #28]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #30]
@2:
.endm

	;;@ Draw a line of a sprite xflipped.
.macro DRAW_SPR_LINE16_XFLIP
	ldr regTile,[regSrc,#4]
	@cmp regTile,#0
	@beq 1f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #2]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #4]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #6]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #8]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #10]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #12]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #14]

@1:
	ldr regTile,[regSrc]
	@cmp regTile,#0
	@beq 2f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #16]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #18]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #20]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #22]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #24]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #26]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #28]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #30]
@2:
.endm

.macro DRAW_LINE16_XFLIP
	DRAW_SPR_LINE16_XFLIP
	add regSrc, regSrc, #8
	add regDst, regDst, #PITCH
.endm
.macro DRAW_LINE16_XYFLIP
	DRAW_SPR_LINE16_XFLIP
	add regSrc, regSrc, #8
	sub regDst, regDst, #PITCH
.endm
.macro DRAW_LINE16_YFLIP
	DRAW_SPR_LINE16
	add regSrc, regSrc, #8
	sub regDst, regDst, #PITCH
.endm
.macro DRAW_LINE16_NORM
	DRAW_SPR_LINE16
	add regSrc, regSrc, #8
	add regDst, regDst, #PITCH
.endm

.macro DRAW_LINE16_XFLIP_ROWS
	DRAW_SPR_LINE16_XFLIP
	add regSrc, regSrc, #8
	add regDst, regDst, #PITCH
	ldr regTemp, [regRows, #4]
	add regDst, regDst, regTemp
	add regRows, regRows, #4
.endm
.macro DRAW_LINE16_XYFLIP_ROWS
	DRAW_SPR_LINE16_XFLIP
	add regSrc, regSrc, #8
	sub regDst, regDst, #PITCH
	ldr regTemp, [regRows, #4]
	add regDst, regDst, regTemp
	add regRows, regRows, #4
.endm
.macro DRAW_LINE16_YFLIP_ROWS
	DRAW_SPR_LINE16
	add regSrc, regSrc, #8
	sub regDst, regDst, #PITCH
	ldr regTemp, [regRows, #4]
	add regDst, regDst, regTemp
	add regRows, regRows, #4
.endm
.macro DRAW_LINE16_NORM_ROWS
	DRAW_SPR_LINE16
	add regSrc, regSrc, #8
	add regDst, regDst, #PITCH
	ldr regTemp, [regRows, #4]
	add regDst, regDst, regTemp
	add regRows, regRows, #4
.endm

	;;@ Draw a line of a sprite.
.macro DRAW_SPR_LINE32
	ldr regTile,[regSrc]
	@cmp regTile,#0
	@beq 1f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #2]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #4]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #6]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #8]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #10]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #12]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #14]

@1:
	ldr regTile,[regSrc,#4]
	@cmp regTile,#0
	@beq 2f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #16]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #18]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #20]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #22]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #24]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #26]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #28]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #30]
@2:
	ldr regTile,[regSrc,#8]
	@cmp regTile,#0
	@beq 3f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #32]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #34]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #36]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #38]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #40]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #42]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #44]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #46]

@3:
	ldr regTile,[regSrc,#12]
	@cmp regTile,#0
	@beq 4f

	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #48]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #50]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #52]
	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #54]

	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #56]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #58]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #60]
	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #62]
@4:
.endm

	;;@ Draw a line of a sprite xflipped.
.macro DRAW_SPR_LINE32_XFLIP
	ldr regTile,[regSrc,#12]
	@cmp regTile,#0
	@beq 1f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #2]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #4]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #6]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #8]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #10]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #12]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #14]

@1:
	ldr regTile,[regSrc,#8]
	@cmp regTile,#0
	@beq 2f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #16]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #18]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #20]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #22]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #24]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #26]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #28]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #30]
@2:
	ldr regTile,[regSrc,#4]
	@cmp regTile,#0
	@beq 3f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #32]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #34]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #36]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #38]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #40]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #42]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #44]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #46]

@3:
	ldr regTile,[regSrc]
	@cmp regTile,#0
	@beq 4f

	ands regPixel,regMask,regTile, lsr #28
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #48]
	ands regPixel,regMask,regTile, lsr #24
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #50]
	ands regPixel,regMask,regTile, lsr #20
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #52]
	ands regPixel,regMask,regTile, lsr #16
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #54]

	ands regPixel,regMask,regTile, lsr #12
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #56]
	ands regPixel,regMask,regTile, lsr #8
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #58]
	ands regPixel,regMask,regTile, lsr #4
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #60]
	ands regPixel,regMask,regTile
	ldrne regColor,[regPal, regPixel, lsl #2]
	strneh regColor,[regDst, #62]
@4:
.endm

.macro DRAW_LINE32_XFLIP
	DRAW_SPR_LINE32_XFLIP
	add regSrc, regSrc, #16
	add regDst, regDst, #PITCH
.endm
.macro DRAW_LINE32_XYFLIP
	DRAW_SPR_LINE32_XFLIP
	add regSrc, regSrc, #16
	sub regDst, regDst, #PITCH
.endm
.macro DRAW_LINE32_YFLIP
	DRAW_SPR_LINE32
	add regSrc, regSrc, #16
	sub regDst, regDst, #PITCH
.endm
.macro DRAW_LINE32_NORM
	DRAW_SPR_LINE32
	add regSrc, regSrc, #16
	add regDst, regDst, #PITCH
.endm

.macro DRAW_TILE8 name
	stmdb sp!, {r3-r9}
	mov regLoop, #8
	mov regMask, #0xF

.lineloop8_\name:
	DRAW_LINE8_\name
	subs regLoop, regLoop, #1
	bne .lineloop8_\name

	ldmia sp!, {r3-r9}
	mov pc,lr               ;@ return
.endm

.macro DRAW_TILE16 name
	stmdb sp!, {r3-r9}
	mov regLoop, #16
	mov regMask, #0xF

.lineloop16_\name:
	DRAW_LINE16_\name
	subs regLoop, regLoop, #1
	bne .lineloop16_\name

	ldmia sp!, {r3-r9}
	mov pc,lr               ;@ return
.endm

.macro DRAW_TILE32 name
	stmdb sp!, {r3-r9}
	mov regLoop, #32
	mov regMask, #0xF

.lineloop32_\name:
	DRAW_LINE32_\name
	subs regLoop, regLoop, #1
	bne .lineloop32_\name

	ldmia sp!, {r3-r9}
	mov pc,lr               ;@ return
.endm

	.align 4
draw_tile_8_xyflip:
	add regDst, regDst, #PITCH * 7
	DRAW_TILE8 XYFLIP

	.align 4
draw_tile_8_xflip:
	DRAW_TILE8 XFLIP

	.align 4
draw_tile_8_yflip:
	add regDst, regDst, #PITCH * 7
	DRAW_TILE8 YFLIP

	.align 4
draw_tile_8_norm:
	DRAW_TILE8 NORM

	.align 4
draw_tile_16_xyflip:
	add regDst, regDst, #PITCH * 15
	DRAW_TILE16 XYFLIP

	.align 4
draw_tile_16_xflip:
	DRAW_TILE16 XFLIP

	.align 4
draw_tile_16_yflip:
	add regDst, regDst, #PITCH * 15
	DRAW_TILE16 YFLIP

	.align 4
draw_tile_16_norm:
	DRAW_TILE16 NORM

	.align 4
draw_tile_rows_xyflip:
	add regDst, regDst, #PITCH * 15
	DRAW_TILE16 XYFLIP_ROWS

	.align 4
draw_tile_rows_xflip:
	DRAW_TILE16 XFLIP_ROWS

	.align 4
draw_tile_rows_yflip:
	add regDst, regDst, #PITCH * 15
	DRAW_TILE16 YFLIP_ROWS

	.align 4
draw_tile_rows_norm:
	DRAW_TILE16 NORM_ROWS

	.align 4
draw_tile_32_xyflip:
	add regDst, regDst, #PITCH * 31
	DRAW_TILE32 XYFLIP

	.align 4
draw_tile_32_xflip:
	DRAW_TILE32 XFLIP

	.align 4
draw_tile_32_yflip:
	add regDst, regDst, #PITCH * 31
	DRAW_TILE32 YFLIP

	.align 4
draw_tile_32_norm:
	DRAW_TILE32 NORM
